### PR TITLE
HRQB 27 - establish integrity checks pattern and Employees check

### PR DESCRIPTION
**Pull request by ghukill**
### Purpose and background context

This PR introduces integrity checks to the HRQBClient for testing data either prepared by tasks or upserted into Quickbase.  These are "runtime" tests in the sense that they assert data quality and integrity during an actual pipeline run. 

This is opposed to unit tests which are helpful for confirming the schema of tasks inputs and outputs, or business logic in the tasks.

These integrity checks are designed to catch data quality issues that may crop up from data changing in the data warehouse or Quickbase, which unit tests could not account for.

If an integrity check fails, the task is considered unsuccessful and details about the failure are reported to Sentry for investigation.

Integrity checks are applied to tasks in the following way:
  1. `HRQBTask` has [new method `integrity_check`](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R110-R116) designed to be used as a decorator
  2. a method in a task is decorated with `HRQBTask.integrity_check` ([example](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-244333a113ecfa98fd1adc7608d4d8ccd9dfa02c6137250a151530ad408870bcR83-R88))
  3. this [registers the method](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R113-R115) as an integrity check at the class level
  4. `HRQBTask` contains a [new method `run_integrity_checks()`](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R118-R134)
  5. both `PandasPickleTask` and `QuickbaseUpsertTask` run `self.run_integrity_checks()` during their `run()` method, passing either the dataframe it would output (`PandasPickleTask` [example](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R169)) or the results of the upsert (`QuickbaseUpsertTask` [example](https://github.com/MITLibraries/hrqb-client/pull/113/files#diff-d86e429c2a92dbd803698c92d15054d84cd0886541acd33a95f26ec5e48d39e6R307)) for checking

This PR is largely to establish the pattern of these checks, but `Employees` does get the first real check asserting that duplicate `MIT ID`'s would not get upserted to the Quickbase `Employees` table.

### How can a reviewer manually see the effects of these changes?

The new unit tests and fixtures do a decent job of outlining how they function.

### Includes new or updated dependencies?
NO

### Changes expectations for external applications?
YES: tasks may now fail, and report to Sentry, if integrity checks fail.

### What are the relevant tickets?
- Include links to Jira Software and/or Jira Service Management tickets here.

### Developer
- [X] All new ENV is documented in README
- [X] All new ENV has been added to staging and production environments
- [X] All related Jira tickets are linked in commit message(s)
- [X] Stakeholder approval has been confirmed (or is not needed)

### Code Reviewer(s)
- [ ] The commit message is clear and follows our guidelines (not just this PR message)
- [ ] There are appropriate tests covering any new functionality
- [ ] The provided documentation is sufficient for understanding any new functionality introduced
- [ ] Any manual tests have been performed **or** provided examples verified
- [ ] New dependencies are appropriate or there were no changes

### CODE REVIEW

File: hrqb/base/task.py [hunk 1]
```diff
@@ -105,6 +106,30 @@ def named_inputs(self) -> dict[str, luigi.Target]:
             for task, target in list(zip(self.deps(), self.input(), strict=True))
         }
 
+    @classmethod
+    def integrity_check(cls, func: Callable) -> Callable:
+        """Decorator used to register integrity check methods from task classes."""
+        if not hasattr(cls, "_integrity_checks"):
+            cls._integrity_checks = set()
+        cls._integrity_checks.add(func.__name__)
+        return func
```
Comment (ehanson8) [inline hrqb/base/task.py:116]: The timing on this PR is impeccable 🙂

File: hrqb/base/task.py [hunk 2]
```diff
@@ -278,6 +278,17 @@ def _normalize_records_for_upsert(self, records_df: pd.DataFrame) -> pd.DataFram
...
+        """Perform Quickbase upsert given a list of record dictionaries."""
+        qbclient = QBClient()
+        table_id = qbclient.get_table_id(self.table_name)
+        upsert_payload = qbclient.prepare_upsert_payload(
+            table_id,
+            records,
+            merge_field=self.merge_field,
+        )
+        return qbclient.upsert_records(upsert_payload)
+
```
Comment (ehanson8) [inline hrqb/base/task.py:291]: This feels like a natural evolution of the `run` method, good update!

File: tests/test_base_task.py [hunk 3]
```diff
@@ -331,3 +332,84 @@ def test_base_pipeline_task_aggregate_upsert_results_upsert_with_errors_noted(
...
+
+def test_base_task_run_integrity_checks_invokes_present_and_skips_absent_checks(
+    pandas_task_with_integrity_checks,
+):
+    """
+    This test simulates another task registering an integrity check that this task does
+    not have.  The task's defined integrity checks are invoked, while this other one is
+    quietly skipped.
+    """
+    pandas_task_with_integrity_checks._integrity_checks.add("i_do_not_have")
```
Comment (ehanson8) [inline tests/test_base_task.py:356]: The tests are clear overall but I don't fully understand how you know `i_do_not_have` isn't run in this test. Is it just that it doesn't raise an error?

File: hrqb/tasks/employees.py [hunk 4]
```diff
@@ -79,6 +80,13 @@ def get_dataframe(self) -> pd.DataFrame:
         }
         return employees_df[fields.keys()].rename(columns=fields)
 
+    @HRQBTask.integrity_check
+    def check_unique_mit_ids(self, output_df: pd.DataFrame) -> None:
+        duplicate_mit_ids = len(output_df[output_df.duplicated("MIT ID", keep=False)])
+        if duplicate_mit_ids > 0:
+            message = f"Found {duplicate_mit_ids} duplicate 'MIT ID' column values"
+            raise ValueError(message)
+
```
Comment (ehanson8) [inline hrqb/tasks/employees.py:89]: Very clear check and test!

**Review by ehanson8 — APPROVED** [review]
Looks good, one question but it's more for clarification than a blocking question

File: hrqb/base/task.py (hunk 1 above)
Comment (ghukill) [reply hrqb/base/task.py:116]: No kidding!  I think the decorators here make sense, given that the integrity check methods will have no naming convention.  But was definitely thinking about Transmog as well.

File: tests/test_base_task.py (hunk 3 above)
Comment (ghukill) [reply tests/test_base_task.py:356]: Nice catch; honing in on a subtle detail here.  Relevant to our decorators conversation in a recent Transmog discussion, when tasks register an integrity check it's at the `HRQBTask` **class** level.  

So, when any given task calls `self._integrity_checks`, it will get back a list of **ALL** integrity check method names, across all tasks.  

This may seem problematic at first, but it's only a name.  So when a task starts to run tests, one of two things happen:
1. it has a method with that name, and it runs its own method
2. it does NOT have a method with that name, quietly skips

This was simulating that another task had registered a check with name `i_do_not_have`, and this test just quietly skips it.  I agree it's not immediately obvious, but I think it really doesn't have any bearing on coding logic: any task is free to define whatever method names it wants for integrity checks.  And, there is no "bleed over" where one task runs another, because it's just by name, not an actual reference to those other methods.

File: hrqb/tasks/employees.py (hunk 4 above)
Comment (ghukill) [reply hrqb/tasks/employees.py:89]: Thanks!  I know there are more to come... this was just the first absolutely lowest fruit to create.  Future ones will be more subtle, e.g. "no annual performance review if job start < 3 months before XX/YY date, and not part of ZZZ union".   

The idea is that these can grow organically as we encounter edge cases, or HR shares that tweaks are needed and we want to make sure we don't regress.

File: tests/test_base_task.py (hunk 3 above)
Comment (ehanson8) [reply tests/test_base_task.py:356]: Got it, thank you for clarifying!

**Review by jonavellecuerdo — APPROVED** [review]
Really cool work here! ✨
